File: solution/725. Split Linked List in Parts/725.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    vector<ListNode*> splitListToParts(ListNode* head, int k) {
        vector<ListNode*> ans(k, nullptr);
        if (!head) return ans;
        
        int n = 0;
        for (auto p = head; p; p = p->next, ++n);
        int len = n / k, remains = n % k;
        
        ListNode *pre = nullptr, *cur = head;
        for (int i = 0; i < k; ++i, --remains) {
            ans[i] = cur;
            for (int j = 0; j < len + (remains > 0); ++j) {
                pre = cur;
                cur = cur->next;
            }
            pre->next = nullptr;
        }
        return ans;
    }
};

File: solution/725. Split Linked List in Parts/725.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:
        n = 0
        p = head
        while p:
            n += 1
            p = p.next
        
        l, r = n // k, n % k
        
        res = [None] * k
        
        tail = None
        p = head
        for i in range(k):
            if not p:
                break
            
            res[i] = p
            
            for _ in range(l):
                tail = p
                p = p.next
            
            if r > 0:
                r -= 1
                tail = p
                p = p.next
            
            if tail:
                tail.next = None
        
        return res
    
